proxy/lib/ketama.c

```diff
@@ -176,17 +176,30 @@ static int ketama_new(lua_State *L) {
         lua_pop(L, 1);
 
         size_t total = 0;
-        for (int x = 1; x < PARTS; x++) {
+        for (int x = 0; x < PARTS; x++) {
             total += partlens[x];
         }
         // add extra space for the repititons.
         char *buf = malloc(total + 16);
 
+        // add 3 more bytes for the delimiters used
+        total = total + 3;
+
         for (int k = 0; k < bucket_size / 4; k++) {
             unsigned char digest[16];
 
+            int num_bytes = 0;
+            int dup_k = k;
+
+            while (dup_k != 0) {
+                dup_k = dup_k / 10;
+                ++num_bytes;
+            }
+            // special case
+            if (k == 0) ++num_bytes;
```

Comment:
Nice catch! The actual bug here is that I added to the total malloc instead of increasing total by 16; the second argument to sprintf doesn't need to be exact, it needs to be as large as it can possibly be. So much simpler if you drop most of this and just total += 16 after the summation and drop the + 16 from malloc?

Also just noticed I've defined "total" twice! Bonus points if the second total gets a name adjustment :)

Reply:
I see, so the idea is to give enough extra space for delimiters and the index without it being accurate? And we expect that it never goes beyond 16 bytes.

Reply:
I probably just wrote 16 without thinking too hard (3 delimiters + null byte + 10-11 characters max for an int, but will realistically not be more than 4-5 without running the system out of RAM). If your calculations went over the 16 that I hard coded we'd have the same issue :)

Could leave a comment as to why 16, or double it to 32, or w/e. doesn't matter much since it's a temporary buffer.